File: app/calculo/secoes/resumo.py
# -*- coding: utf-8 -*-
"""Sumário executivo — porte do cálculo de renderSumario() (app.js)."""
from ..dre import dre_agg
from ..metas import metas_calc
from ..numeros import soma
from ..sazonalidade import titulo_crescimento
from ..vendas import agregar


def _ids_linha(r):
    """Identidade de relação: arquiteto quando a venda tem arquiteto, senão o cliente."""
    return ['a%s' % x[0] for x in r[16]] if r[16] else ['c%s' % r[5]]


def _identidades_unicas(DATA, a, b):
    pos = {}
    for r in DATA['rows']:
        if r[0] < a or r[0] > b:
            continue
        for k in _ids_linha(r):
            pos[k] = pos.get(k, 0) + r[1]
    return sum(1 for v in pos.values() if v > 0)


def calcular(C, params=None):
    DATA, DRE, DPNL = C.blob('DATA'), C.blob('DRE'), C.blob('DPNL')
    fim = C.maxym_vendas
    y1, y0 = C.ano_v, C.ano_v - 1
    i1 = y1 * 100 + 1
    A = agregar(DATA, 200001, 300000)
    c = A['comp']
    rep = c['pf'] + c['dsr'] + c['prem'] + c['pj'] + c['ger'] + c['cur'] + c['rt'] + c['roy']
    top10 = soma(x['v'] for x in A['clis'][:10])
    dtop = soma(x['roy'] for x in A['desg'][:2])
    dtot = soma(x['roy'] for x in A['desg'])

    antes, no_ano = set(), set()
    for r in DATA['rows']:
        ids = _ids_linha(r)
        if r[0] < i1:
            antes.update(ids)
        elif r[0] <= fim:
            no_ano.update(ids)
    novos = sum(1 for k in no_ano if k not in antes)
    H = agregar(DATA, i1, fim)

    g26 = dre_agg(DPNL, 'CONSOLIDADO', C.ano_d * 100 + 1, C.maxym_dre) if DPNL else None
    fat25ytd = dre_agg(DPNL, 'CONSOLIDADO', (C.ano_d - 1) * 100 + 1, C.maxym_dre - 100)['receita_bruta'] if DPNL else 0
    MP = metas_calc(C)
    base = {'maxym': fim, 'minym': C.minym_vendas, 'Y0': y0, 'Y1': y1}
    return {
        'resumo.abertura': dict(base, n_linhas=len(DATA['rows'])),
        'rs-hist': dict(base, total=A['total'], peds=A['peds'], ticket=A['ticket'],
                        rel_unicas=_identidades_unicas(DATA, 200001, 300000), n_designers=len(A['desg']),
                        dtop=dtop, dtot=dtot, ano_d_ant=C.ano_d - 1,
                        fat_ano_ant=(DRE.get(str(C.ano_d - 1)) or {}).get('annual') if DRE else None,
                        carteira=DATA['carteira']['total'], rep=rep, top10=top10),
        'rs-ytd': dict(base, total=H['total'], peds=H['peds'], ticket=H['ticket'],
                       orcado=MP['metaAteAgora'] if MP else None, realizado=MP['realAteAgora'] if MP else None,
                       dre={'fat': g26['receita_bruta'], 'fat_ant': fat25ytd, 'ebitda': g26['ebitda'], 'eb_pct': g26['eb_pct'],
                            'ebit': g26['ebitda'] + g26['deprec'], 'rl': g26['receita_liquida'] or 1,
                            'ano_d_ant': C.ano_d - 1} if g26 else None,
                       rel_unicas=_identidades_unicas(DATA, i1, fim), novos=novos),
        'rs-leitura': dict(base, titulo=titulo_crescimento(C), ano_ini=C.ano_ini_serie,
                           venda_ini=agregar(DATA, C.ano_ini_serie * 100 + 1, C.ano_ini_serie * 100 + 12)['total'],
                           venda_ant=agregar(DATA, y0 * 100 + 1, y0 * 100 + 12)['total'], venda_ano=H['total'],
                           orcado=MP['metaAteAgora'] if MP else None, realizado=MP['realAteAgora'] if MP else None,
                           dtop=dtop, dtot=dtot, rt=c['rt'], total=A['total'], top10=top10),
    }

File: app/static/relatorio/secoes/resumo.js
/* Sumário executivo — desenho de renderSumario() do Kit, alimentado pelo servidor. */
'use strict';
window.DESENHO=window.DESENHO||{};
DESENHO.resumo=function(P){
  let s='';
  const ymLabExt=ym=>['janeiro','fevereiro','março','abril','maio','junho','julho','agosto','setembro','outubro','novembro','dezembro'][(ym%100)-1]+'/'+Math.floor(ym/100);
  const base=P['resumo.abertura']||P['rs-hist']||P['rs-ytd']||P['rs-leitura'];
  if(!base) return s;
  const mLabV=MES[(base.maxym%100)-1], ymLabV=ymLab(base.maxym), Y1=base.Y1, Y0=base.Y0;
  const A=P['resumo.abertura'];
  if(A) s+=`<div class="sec-head"><div class="kick">+55 Design · base atualizada até ${ymLabV}</div><h2>Sumário executivo de vendas</h2>
   <p class="lead">Panorama em duas leituras: o <b>histórico completo</b> desde ${ymLabAno(A.minym)} e um recorte específico de <b>${Y1} (Jan–${mLabV})</b>. Base comercial de ${nf(A.n_linhas)} linhas; faturamento e resultado pela DRE (abas REAL). Filtros de período na seção “Análise por período” e na DRE.</p></div>`;
  const H=P['rs-hist'];
  if(H){
    s+=H3('Histórico · '+ymLabExt(H.minym)+' – '+mLabV.toLowerCase()+'/'+Y1,'','rs-hist');
    s+='<div class="kpis">'
     +kpi('Venda contratada · total',mi(H.total),ymLabAno(H.minym)+'–'+ymLabV,'ok')
     +kpi('Pedidos',nf(H.peds),'ticket médio '+mi(H.ticket))
     +kpi('Relações únicas',nf(H.rel_unicas),'arquiteto, ou cliente sem arquiteto — compra líquida positiva')
     +kpi('Designers licenciados',nf(H.n_designers),'2 maiores = '+pct(H.dtop/H.dtot,0)+' dos royalties')
     +kpi('Faturamento '+H.ano_d_ant+' · DRE',H.fat_ano_ant!=null?mi(H.fat_ano_ant):'—','Receita Bruta Consolidada')
     +kpi('Carteira em aberto',mi(H.carteira),'pos. '+ymLabV,'warn')
     +kpi('Repasses e comissões',mi(H.rep),pct(H.rep/H.total,1)+' das vendas')
     +kpi('Concentração Top 10',pct(H.top10/H.total,0),'dos clientes na venda')
     +'</div>';
  }
  const Y=P['rs-ytd'];
  if(Y){
    const hasMeta=Y.orcado!=null, D=Y.dre;
    s+=H3(Y1+' · Janeiro a '+mLabV,'','rs-ytd');
    s+='<div class="kpis">'
     +kpi('Venda contratada · '+Y1,mi(Y.total),hasMeta?deltaHtml(Y.realizado/Y.orcado-1)+' vs orçado':'jan–'+ymLabV)
     +(D?kpi('Faturamento · DRE',mi(D.fat),deltaHtml(D.fat/D.fat_ant-1)+' vs '+D.ano_d_ant+' YTD'):'')
     +(D?kpi('EBITDA · DRE',mi(D.ebitda),pct(D.eb_pct,1)+' da ROL',D.ebitda>=0?'ok':'warn'):'')
     +(D?kpi('EBIT · DRE',mi(D.ebit),pct(D.ebit/D.rl,1)+' da ROL — antes do efeito financeiro',D.ebit>=0?'ok':'warn'):'')
     +kpi('Pedidos · '+Y1,nf(Y.peds),'ticket médio '+mi(Y.ticket))
     +kpi('Ticket médio · '+Y1,mi(Y.ticket),'por pedido')
     +kpi('Relações únicas · '+Y1,nf(Y.rel_unicas),'arquiteto, ou cliente sem arquiteto — compra líquida positiva')
     +kpi('Novas relações · '+Y1,nf(Y.novos),'arquiteto ou cliente, sem compra antes de '+Y1,'ok')
     +'</div>';
  }
  const L=P['rs-leitura'];
  if(L){
    const hasMeta=L.orcado!=null;
    const prov=[
     [L.titulo,'Venda contratada saltou de '+mi(L.venda_ini)+' ('+L.ano_ini+') para '+mi(L.venda_ant)+' ('+Y0+'). '+Y1+' (Jan–'+mLabV+') soma '+mi(L.venda_ano)+(hasMeta?(L.realizado<L.orcado?', abaixo do orçado ('+mi(L.orcado)+').':', acima do orçado ('+mi(L.orcado)+').'):'.')],
     ['Concentração de design','Os dois maiores estúdios concentram '+pct(L.dtop/L.dtot,0)+' dos royalties pagos ('+mi(L.dtot)+' no total).'],
     ['RT a arquitetos é o maior repasse','O repasse técnico a arquitetos soma '+mi(L.rt)+' ('+pct(L.rt/L.total,1)+' das vendas), acima até dos royalties de designer.'],
     ['Baixa concentração de clientes','Top 10 clientes = '+pct(L.top10/L.total,0)+' das vendas — carteira pulverizada, com espaço para recompra e indicação.'],
    ];
    s+='<h3 data-blk="rs-leitura">Leitura executiva</h3><div class="prov">'+prov.map((p,i)=>`<div class="pv"><div class="pv-n">${i+1}</div><div><b>${esc(p[0])}</b><p>${p[1]}</p></div></div>`).join('')+'</div>';
  }
  return s;
};
